File: ios/Servantana/Sources/ViewModels/ChatViewModel.swift
import Foundation
import Combine

@MainActor
class ChatViewModel: ObservableObject {
    @Published var messages: [Message] = []
    @Published var otherUserName = "Chat"
    @Published var isLoading = false
    @Published var isSending = false
    @Published var error: String?
    @Published var isPolling = false

    private let userId: String
    private var pollingTask: Task<Void, Never>?
    private var cancellables = Set<AnyCancellable>()
    private let pollingInterval: TimeInterval = 3.0 // Poll every 3 seconds

    init(userId: String) {
        self.userId = userId
        Task {
            await loadMessages()
        }
        setupNotificationObserver()
    }

    deinit {
        pollingTask?.cancel()
    }

    // MARK: - Message Loading

    func loadMessages() async {
        isLoading = true
        error = nil

        do {
            let response = try await APIClient.shared.getMessages(conversationId: userId)
            let newMessages = response.messages.sorted { $0.timestamp < $1.timestamp }

            // Only update if there are new messages to avoid UI flicker
            if newMessages.count != messages.count ||
               newMessages.last?.id != messages.last?.id {
                messages = newMessages
            }
        } catch {
            self.error = error.localizedDescription
        }

        isLoading = false
    }

    func sendMessage(_ content: String) async {
        isSending = true

        do {
            let request = SendMessageRequest(receiverId: userId, content: content, bookingId: nil)
            let response = try await APIClient.shared.sendMessage(request)
            messages.append(response.message)
        } catch {
            self.error = error.localizedDescription
        }

        isSending = false
    }

    // MARK: - Real-time Polling

    /// Start polling for new messages when chat view is visible
    func startPolling() {
        guard pollingTask == nil else { return }
        isPolling = true

        pollingTask = Task { [weak self] in
            while !Task.isCancelled {
                try? await Task.sleep(nanoseconds: UInt64(self?.pollingInterval ?? 3.0) * 1_000_000_000)
                guard !Task.isCancelled else { break }
                await self?.refreshMessages()
            }
        }
    }

    /// Stop polling when chat view disappears
    func stopPolling() {
        pollingTask?.cancel()
        pollingTask = nil
        isPolling = false
    }

    /// Refresh messages without showing loading state (for background updates)
    private func refreshMessages() async {
        do {
            let response = try await APIClient.shared.getMessages(conversationId: userId)
            let newMessages = response.messages.sorted { $0.timestamp < $1.timestamp }

            // Check if there are actual new messages
            if let lastNew = newMessages.last,
               let lastCurrent = messages.last,
               lastNew.id != lastCurrent.id || newMessages.count > messages.count {
                messages = newMessages
            }
        } catch {
            // Silently fail on polling errors
            print("Polling error: \(error)")
        }
    }

    // MARK: - Push Notification Handling

    private func setupNotificationObserver() {
        NotificationCenter.default.publisher(for: .pushNotificationTapped)
            .receive(on: DispatchQueue.main)
            .sink { [weak self] notification in
                guard let self = self,
                      let userInfo = notification.userInfo,
                      let pushNotification = userInfo["notification"] as? PushNotification,
                      pushNotification.type == "MESSAGE_RECEIVED" else { return }

                // Refresh messages when a message notification is tapped
                Task {
                    await self.loadMessages()
                }
            }
            .store(in: &cancellables)
    }
}

File: android/app/src/main/java/com/servantana/app/ui/screens/messages/ChatViewModel.kt
package com.servantana.app.ui.screens.messages

import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.servantana.app.data.model.Message
import com.servantana.app.data.model.User
import com.servantana.app.data.repository.MessageRepository
import com.servantana.app.data.repository.WorkerRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import java.util.UUID
import javax.inject.Inject

data class ChatUiState(
    val isLoading: Boolean = false,
    val otherUser: User? = null,
    val messages: List<Message> = emptyList(),
    val isSending: Boolean = false,
    val error: String? = null,
    val isPolling: Boolean = false
)

@HiltViewModel
class ChatViewModel @Inject constructor(
    private val messageRepository: MessageRepository,
    private val workerRepository: WorkerRepository,
    savedStateHandle: SavedStateHandle
) : ViewModel() {

    private val userId: String = checkNotNull(savedStateHandle["userId"])

    private val _uiState = MutableStateFlow(ChatUiState())
    val uiState: StateFlow<ChatUiState> = _uiState.asStateFlow()

    private var pollingJob: Job? = null
    private val pollingIntervalMs = 3000L // 3 seconds

    init {
        loadChat()
    }

    override fun onCleared() {
        super.onCleared()
        stopPolling()
    }

    private fun loadChat() {
        viewModelScope.launch {
            _uiState.update { it.copy(isLoading = true, error = null) }

            // Load other user info
            workerRepository.getWorker(userId)
                .onSuccess { worker ->
                    _uiState.update {
                        it.copy(
                            otherUser = User(
                                id = worker.id,
                                email = worker.email,
                                firstName = worker.firstName,
                                lastName = worker.lastName,
                                avatar = worker.avatar,
                                role = worker.role
                            )
                        )
                    }
                }

            // Load messages
            messageRepository.getMessages(userId)
                .onSuccess { messages ->
                    _uiState.update {
                        it.copy(
                            isLoading = false,
                            messages = messages.sortedBy { msg -> msg.timestamp }
                        )
                    }
                }
                .onFailure { exception ->
                    _uiState.update {
                        it.copy(
                            isLoading = false,
                            error = exception.message ?: "Failed to load messages"
                        )
                    }
                }
        }
    }

    fun sendMessage(content: String) {
        if (content.isBlank()) return

        viewModelScope.launch {
            _uiState.update { it.copy(isSending = true) }

            // Optimistically add message to UI
            val tempMessage = Message(
                id = UUID.randomUUID().toString(),
                senderId = "currentUser", // Will be replaced by actual user ID
                receiverId = userId,
                content = content.trim(),
                timestamp = System.currentTimeMillis(),
                isRead = false
            )

            _uiState.update {
                it.copy(messages = it.messages + tempMessage)
            }

            messageRepository.sendMessage(userId, content.trim())
                .onSuccess { message ->
                    // Replace temp message with actual message
                    _uiState.update {
                        it.copy(
                            isSending = false,
                            messages = it.messages.dropLast(1) + message
                        )
                    }
                }
                .onFailure { exception ->
                    // Remove temp message and show error
                    _uiState.update {
                        it.copy(
                            isSending = false,
                            messages = it.messages.dropLast(1),
                            error = exception.message ?: "Failed to send message"
                        )
                    }
                }
        }
    }

    fun refresh() {
        loadChat()
    }

    fun clearError() {
        _uiState.update { it.copy(error = null) }
    }

    // MARK: - Real-time Polling

    /**
     * Start polling for new messages when chat view is visible
     */
    fun startPolling() {
        if (pollingJob != null) return

        _uiState.update { it.copy(isPolling = true) }

        pollingJob = viewModelScope.launch {
            while (isActive) {
                delay(pollingIntervalMs)
                refreshMessages()
            }
        }
    }

    /**
     * Stop polling when chat view disappears
     */
    fun stopPolling() {
        pollingJob?.cancel()
        pollingJob = null
        _uiState.update { it.copy(isPolling = false) }
    }

    /**
     * Refresh messages without showing loading state (for background updates)
     */
    private suspend fun refreshMessages() {
        messageRepository.getMessages(userId)
            .onSuccess { newMessages ->
                val sortedMessages = newMessages.sortedBy { it.timestamp }
                val currentMessages = _uiState.value.messages

                // Only update if there are new messages
                if (sortedMessages.size != currentMessages.size ||
                    sortedMessages.lastOrNull()?.id != currentMessages.lastOrNull()?.id) {
                    _uiState.update { it.copy(messages = sortedMessages) }
                }
            }
            // Silently ignore polling errors
    }
}
